### PR TITLE
Add ability to revoke consent sessions

### PR DESCRIPTION
**Pull request by Sheikah45**

### CODE REVIEW

**Comment by Sheikah45** [comment]
Going to simplify this to having an access point to revoking all consent sessions for now

**Comment by Sheikah45** [comment]
Right now it is a minimum viable access point that the moderator client can use to revoke consent sessions when someone is banned

**Comment by Sheikah45** [comment]
Still need to add integration tests

File: src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt [hunk 1]
```diff
@@ -114,4 +119,14 @@ class OAuthController(
         }.flatMap { redirectUrl ->
             redirect(response, redirectUrl)
         }
+
+    @PostMapping("revokeSession")
+    @PreAuthorize("hasRole('" + FafRole.ADMIN_ACCOUNT_BAN + "') + hasRole('" + OAuthScope.ADMINISTRATIVE_ACTION + "')")
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:124]: hasRole won't work on OAuth scopes.
Also this is Kotlin. If you make it const val string, you don't need to concat the string, just user `hasRole('${fafRole.ADMIN_ACCOUNT_BAN}')`

File: src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt [hunk 2]
```diff
@@ -114,4 +119,14 @@ class OAuthController(
...
+    fun revokeSession(
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:125]: Sessions might be a misleading term in all cases. revokeConsent would be correct from the hydra perspective or revokeRefreshToken from client perspective.

File: src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt [hunk 3]
```diff
@@ -14,21 +14,25 @@ enum class BanLevel {
 }
 
 @Table("ban")
-data class Ban(
+data class BanInfo(
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt:17]: Why renamed?

File: src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt [hunk 4]
```diff
@@ -13,15 +13,20 @@ data class FafAuthority(val role: String) : GrantedAuthority {
...
+
 /**
- * Extract the FAF roles that we put into the ext->roles array of the access token.
+ * Extract the FAF roles that we put into the ext->roles array of the access token and the scopes.
  */
 @Component
 class FafJwtAuthenticationConverter : Converter<Jwt, Flux<GrantedAuthority>> {
     override fun convert(jwt: Jwt): Flux<GrantedAuthority> {
         val ext: JSONObject? = jwt.claims["ext"] as? JSONObject
+        val scopes: JSONArray = jwt.claims["scp"] as JSONArray
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt:27]: you don't need the variable type declarations if you use `as ...`

File: src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt [hunk 5]
```diff
@@ -13,15 +13,20 @@ data class FafAuthority(val role: String) : GrantedAuthority {
     override fun getAuthority() = "ROLE_$role"
 }
 
+data class FafScope(val role: String) : GrantedAuthority {
+    override fun getAuthority() = "SCOPE_$role"
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt:17]: You have a const for SCOPE already. Also maybe move the scope class there

File: src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
+/**
```
Comment (Brutus5000) [inline src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt:1]: Wrong package, wrong preface text.

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt (hunk 3 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt:17]: To make it consistent with how it is named in the api and the java commons

File: src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt (hunk 6 above)
Comment (Sheikah45) [reply src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt:1]: Yep

File: src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt (hunk 5 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt:17]: Still not used to being able to reasonably put multiple classes in a single file

File: src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt (hunk 4 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt:27]: what is this magic?

File: src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt (hunk 2 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:125]: makes sense

File: src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt (hunk 1 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:124]: more magic

File: src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt (hunk 6 above)
Comment (Sheikah45) [reply src/main/kotlin/sh/ory/hydra/model/RevokeSessionsRequest.kt]: Done

File: src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt (hunk 5 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/security/FafJwtAuthenticationConverter.kt]: Refactored

**Comment by Sheikah45** [comment]
Should be good to go now

**Comment by Sheikah45** [comment]
Tested it locally and seemed to work

File: src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt (hunk 3 above)
Comment (Brutus5000) [reply src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt:17]: na BanInfo is a horrible name and it is due to historic reason. let's keep ban please

File: src/main/kotlin/com/faforever/userservice/hydra/HydraService.kt [hunk 7]
```diff
@@ -155,4 +155,18 @@ class HydraService(
             .bodyValue(error)
             .retrieve()
             .bodyToMono(RedirectResponse::class.java)
+
+    fun revokeAllSessions(revokeRefreshTokensRequest: RevokeRefreshTokensRequest): Mono<RedirectResponse> {
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/hydra/HydraService.kt:159]: rename Sessions in method name too please

Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/hydra/HydraService.kt]: DOne

File: src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt (hunk 3 above)
Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/domain/BanInfo.kt]: Got it

File: src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt [hunk 8]
```diff
@@ -134,4 +139,18 @@ class OAuthController(
         }.flatMap { redirectUrl ->
             redirect(response, redirectUrl)
         }
+
+    @PostMapping("revokeTokens")
```
Comment (Brutus5000) [inline src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:143]: Do we not need to expose this endpoint in the SecurityConfig?

Comment (Sheikah45) [reply src/main/kotlin/com/faforever/userservice/controller/OAuthController.kt:143]: no it is covered by /**

**Review by Brutus5000 — APPROVED** [review]

**Pull request merged**